lib/datadog/core/utils/at_fork_monkey_patch.rb

```diff
@@ -50,40 +51,42 @@ def fork
             # If a block is provided, use the wrapped version.
             result = child_block.nil? ? super : super(&child_block)
 
-            # Trigger correct callbacks depending on whether we're in the parent or child.
+            # When fork gets called without a block, it returns twice:
             # If we're in the fork, result = nil: trigger child callbacks.
-            # If we're in the parent, result = fork PID: trigger parent callbacks.
+            # If we're in the parent, result = pid: we do nothing.
+            # (If it gets called with a block, it only returns on the parent)
             datadog_at_fork_blocks[:child].each(&:call) if result.nil? && datadog_at_fork_blocks.key?(:child)
 
-            # Return PID from #fork
             result
           end
 
-          def at_fork(stage, &block)
-            raise ArgumentError, 'Bad \'stage\' for ::at_fork' unless stage == :child
+          # NOTE: You probably want to wrap any calls to datadog_at_fork with a OnlyOnce so as to not re-register
+          #       the same block/behavior more than once.
+          def datadog_at_fork(stage, &block)
+            raise ArgumentError, 'Bad \'stage\' for ::datadog_at_fork' unless stage == :child
 
-            datadog_at_fork_blocks[stage] = [] unless datadog_at_fork_blocks.key?(stage)
+            datadog_at_fork_blocks[stage] ||= []
             datadog_at_fork_blocks[stage] << block
+
+            nil
           end
 
           module_function
 
           def datadog_at_fork_blocks
-            # Blocks should be shared across all users of this module,
+            # Blocks are shared across all users of this module,
             # e.g. Process#fork, Kernel#fork, etc. should all invoke the same callbacks.
-            # rubocop:disable Style/ClassVars
-            @@datadog_at_fork_blocks ||= {}
-            # rubocop:enable Style/ClassVars
+            @@datadog_at_fork_blocks ||= {} # rubocop:disable Style/ClassVars
```

Comment:
I feel like this should be protected by a mutex?

Reply:
Yeah, it's an excellent point. When this code runs the first time we shouldn't see concurrent forking, but "shouldn't" is not "won't".

Since we want to ship some of these changes as soon as possible, I've gone ahead and created https://github.com/DataDog/dd-trace-rb/issues/3831 and I'll come back to refactor this in a few days :)